### PR TITLE
FIX: Re-organize FreeSurfer stages to avoid duplication and races

### PR DESCRIPTION
**Pull request by effigies**
Addresses poldracklab/fmriprep#1812.

Do we run a full FreeSurfer in our tests?

### CODE REVIEW

**Comment by oesteban** [comment]
> Do we run a full FreeSurfer in our tests?

No, we don't.

I believe we need to add that `validate=False` to the pybids' `build_path` calls.

**Comment by oesteban** [comment]
Although niworkflows is 0.10.4 (should contain the patch), it seems that pybids is not correct:
```
Successfully built niworkflows sphinx-argparse duecredit datalad psutil transforms3d humanize wrapt jsmin PyGithub pandocfilters pyrsistent
ERROR: niworkflows 0.10.4 has requirement pybids>=0.9.4, but you'll have pybids 0.9.3 which is incompatible.
Installing collected packages: filelock, etelemetry, nipype, transforms3d, niworkflows, lockfile, coverage, docutils, sphinxcontrib-htmlhelp, alabaster, imagesize, sphinxcontrib-jsmath, snowballstemmer, Pygments, sphinxcontrib-serializinghtml, sphinxcontrib-applehelp, sphinxcontrib-qthelp, babel, sphinxcontrib-devhelp, sphinx, sphinx-argparse, sphinx-rtd-theme, codecov, atomicwrites, more-itertools, zipp, importlib-metadata, pluggy, py, wcwidth, pytest, ipython-genutils, pyrsistent, jsonschema, traitlets, jupyter-core, nbformat, defusedxml, testpath, mistune, pandocfilters, webencodings, bleach, entrypoints, nbconvert, nbsphinx, citeproc-py, duecredit, mock, appdirs, smmap2, gitdb2, GitPython, iso8601, humanize, monotonic, fasteners, patool, wrapt, boto, jeepney, secretstorage, keyring, keyrings.alt, msgpack, jsmin, deprecated, pyjwt, PyGithub, whoosh, datalad, psutil, smriprep
  Found existing installation: nipype 1.2.1
    Uninstalling nipype-1.2.1:
      Successfully uninstalled nipype-1.2.1
  Found existing installation: niworkflows 0.10.2
    Uninstalling niworkflows-0.10.2:
      Successfully uninstalled niworkflows-0.10.2
Successfully installed GitPython-3.0.3 PyGithub-1.43.8 Pygments-2.4.2 alabaster-0.7.12 appdirs-1.4.3 atomicwrites-1.3.0 babel-2.7.0 bleach-3.1.0 boto-2.49.0 citeproc-py-0.4.0 codecov-2.0.15 coverage-4.5.4 datalad-0.11.7 defusedxml-0.6.0 deprecated-1.2.6 docutils-0.15.2 duecredit-0.7.0 entrypoints-0.3 etelemetry-0.1.2 fasteners-0.15 filelock-3.0.12 gitdb2-2.0.6 humanize-0.5.1 imagesize-1.1.0 importlib-metadata-0.23 ipython-genutils-0.2.0 iso8601-0.1.12 jeepney-0.4.1 jsmin-2.2.2 jsonschema-3.1.1 jupyter-core-4.6.0 keyring-19.2.0 keyrings.alt-3.1.1 lockfile-0.12.2 mistune-0.8.4 mock-3.0.5 monotonic-1.5 more-itertools-7.2.0 msgpack-0.6.2 nbconvert-5.6.0 nbformat-4.4.0 nbsphinx-0.4.3 nipype-1.2.3 niworkflows-0.10.4 pandocfilters-1.4.2 patool-1.12 pluggy-0.13.0 psutil-5.6.3 py-1.8.0 pyjwt-1.7.1 pyrsistent-0.15.4 pytest-5.2.1 secretstorage-3.1.1 smmap2-2.0.5 smriprep-0+unknown snowballstemmer-2.0.0 sphinx-2.2.0 sphinx-argparse-0.2.5 sphinx-rtd-theme-0.4.3 sphinxcontrib-applehelp-1.0.1 sphinxcontrib-devhelp-1.0.1 sphinxcontrib-htmlhelp-1.0.2 sphinxcontrib-jsmath-1.0.1 sphinxcontrib-qthelp-1.0.2 sphinxcontrib-serializinghtml-1.1.3 testpath-0.4.2 traitlets-4.3.3 transforms3d-0.3.1 wcwidth-0.1.7 webencodings-0.5.1 whoosh-2.7.4 wrapt-1.11.2 zipp-0.6.0
```

File: smriprep/workflows/surfaces.py [hunk 1]
```diff
@@ -323,28 +329,44 @@ def init_autorecon_resume_wf(omp_nthreads, name='autorecon_resume_wf'):
...
         iterfield='hemi', n_procs=omp_nthreads, mem_gb=5,
         name='autorecon_surfs')
     autorecon_surfs.inputs.hemi = ['lh', 'rh']
     autorecon_surfs.interface._always_run = True
 
-    autorecon3 = pe.MapNode(
-        fs.ReconAll(directive='autorecon3', openmp=omp_nthreads),
+    # -cortribbon is a prerequisite for -parcstats, -parcstats2, -parcstats3
+    cortribbon = pe.Node(ReconAll(directive=Undefined,
+                                  steps=['cortribbon']), name='cortribbon')
```
Comment (effigies) [inline smriprep/workflows/surfaces.py]: This task motivates the updated interface. If the output of a given step specified in `steps` exists, and its inputs have not been updated, then the step will be skipped. It is analogous to the `directive` input, which is a shorthand for many steps.

File: smriprep/interfaces/freesurfer.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+        # Mix of autorecon2 and autorecon3 steps
+        'autorecon-pial',
+        'autorecon-hemi',
+        # Not "multi-stage flags"
+        'localGI',
+        'qcache',
+        argstr='-%s',
+        desc='process directive',
+        usedefault=True,
+        xor=['steps'],
```
Comment (effigies) [inline smriprep/interfaces/freesurfer.py]: Only change from the original trait.

File: smriprep/interfaces/freesurfer.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+            'parcstats3',
+            'label-exvivo-ec',
+            # autorecon vol
+            'cortribbon',
+            'hyporelabel',  # 6.0
+            'segstats',
+            'aparc2aseg',
+            'apas2aseg',    # 6.0
+            'wmparc',
+            'balabels',
```
Comment (effigies) [inline smriprep/interfaces/freesurfer.py]: This is a full list of steps that are accepted in FreeSurfer 5.3.0 or 6.0.x.

File: smriprep/interfaces/freesurfer.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+    input_spec = ReconAllInputSpec
+
+    def _is_resuming(self):
+        subjects_dir = self.inputs.subjects_dir
+        if not isdefined(subjects_dir):
+            subjects_dir = self._gen_subjects_dir()
+        if os.path.isdir(
+                os.path.join(subjects_dir, self.inputs.subject_id, 'mri')):
+            return True
+        return False
```
Comment (effigies) [inline smriprep/interfaces/freesurfer.py]: I don't think this changed... I think I just failed to remove it. Will do before merge, but shouldn't affect review.

File: smriprep/interfaces/freesurfer.py [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+        if not isdefined(subjects_dir):
+            subjects_dir = self._gen_subjects_dir()
+
+        # Check only relevant steps
+        directive = self.inputs.directive
+        if not isdefined(directive):
+            if isdefined(self.inputs.steps):
+                steps = [step for step in self._steps if step[0] in self.inputs.steps]
+            else:
+                steps = []
```
Comment (effigies) [inline smriprep/interfaces/freesurfer.py]: This block is changed from upstream.

File: smriprep/interfaces/freesurfer.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
```
Comment (oesteban) [inline smriprep/interfaces/freesurfer.py]:
```suggestion
            steps = []
            if isdefined(self.inputs.steps):
                steps = [step for step in self._steps if step[0] in self.inputs.steps]
```

File: smriprep/workflows/surfaces.py [hunk 7]
```diff
@@ -355,16 +377,20 @@ def _dedup(in_list):
         return vals.pop()
 
     workflow.connect([
-        (inputnode, autorecon3, [('use_T2', 'use_T2'),
-                                 ('use_FLAIR', 'use_FLAIR')]),
+        (inputnode, autorecon_surfs, [('use_T2', 'use_T2'),
+                                      ('use_FLAIR', 'use_FLAIR')]),
```
Comment (oesteban) [inline smriprep/workflows/surfaces.py]: Do you think this PR will help with #33 in any way? Is there any dataset with FLAIR images on OpenNeuro that we can use to test?

Comment (effigies) [reply smriprep/workflows/surfaces.py]: No, there are no changes to how FLAIRs are handled.

**Pull request merged**